src/test/app/Wasm_test.cpp

```diff
@@ -695,6 +695,1089 @@ struct Wasm_test : public beast::unit_test::suite
+        Env env(*this);
+        auto& engine = WasmEngine::instance();
+
+        // Test gas cost accuracy and consistency
+        {
+            auto const wasmStr = boost::algorithm::unhex(fibWasmHex);
+            Bytes const wasm(wasmStr.begin(), wasmStr.end());
+
+            // Test consistent gas costs for same operations
+            std::vector<int32_t> fibInputs = {
+                5, 5, 5};  // Same input multiple times
+            std::vector<int64_t> costs;
+
+            for (auto input : fibInputs)
+            {
+                auto re = engine.run(wasm, "fib", wasmParams(input));
+                if (BEAST_EXPECT(re.has_value()))
+                {
+                    BEAST_EXPECTS(re->result == 5, std::to_string(re->result));
+                    costs.push_back(re->cost);
+                }
+            }
+
+            // All costs should be identical for identical operations
+            if (costs.size() >= 2)
+            {
+                BEAST_EXPECTS(
+                    costs[0] == costs[1],
+                    std::to_string(costs[0]) +
+                        " != " + std::to_string(costs[1]));
+            }
+        }
+
+        // Test gas cost scaling for different computation sizes
+        {
+            auto const wasmStr = boost::algorithm::unhex(fibWasmHex);
+            Bytes const wasm(wasmStr.begin(), wasmStr.end());
+
+            struct FibCostTest
+            {
+                int32_t input;
+                int32_t expectedResult;
+                int64_t expectedCost;
+            };
+
+            std::vector<FibCostTest> tests = {
+                {5, 5, -1},    // Will fill in actual cost
+                {10, 55, -1},  // Should be higher cost
+                {15, 610, -1}  // Should be even higher
+            };
+
+            for (auto& test : tests)
+            {
+                auto re = engine.run(wasm, "fib", wasmParams(test.input));
+                if (BEAST_EXPECT(re.has_value()))
+                {
+                    BEAST_EXPECTS(
+                        re->result == test.expectedResult,
+                        std::to_string(re->result));
+                    test.expectedCost = re->cost;
+                }
+            }
+
+            // Verify that costs scale appropriately (higher input = higher
+            // cost)
+            if (tests.size() >= 3 && tests[0].expectedCost > 0 &&
+                tests[1].expectedCost > 0 && tests[2].expectedCost > 0)
+            {
+                BEAST_EXPECTS(
+                    tests[0].expectedCost < tests[1].expectedCost,
+                    "fib(5) cost should be less than fib(10) cost");
+                BEAST_EXPECTS(
+                    tests[1].expectedCost < tests[2].expectedCost,
+                    "fib(10) cost should be less than fib(15) cost");
+            }
+        }
+
+        // Test large module handling performance
+        {
+            // Use the largest available WASM module for testing
```

Comment:
We can't put any time related tests, as they must run only on controlled environment.